Add date_sub Spark function

zwan5 opened the pull request:
Add date_sub in sparksql functions

zwan5 commented:
Close it temporarily.

zwan5 commented:
Reopen. Thanks.

mbasmanova reviewed (commented):
@zwan5 Thank you for the contribution. A few rules to follow when adding functions:

1- PR title should be "Add xxx Spark function"
2- PR description should not repeat PR title. Either leave empty or provide non-trivial description of the function behavior or implementation.

mbasmanova commented on velox/docs/functions/spark/datetime.rst:
```diff
@@ -7,6 +7,10 @@ Convenience Extraction Functions
 
+.. spark:function:: date_sub(start_date, num_days) -> date
+
+    Returns the date that is num_days before start_date.
```
Would you clarify if num_days can be negative?

mbasmanova commented on velox/functions/sparksql/DateTimeFunctions.h:
```diff
@@ -209,4 +209,34 @@ struct LastDayFunction {
+      const int32_t value) {
+    int32_t subValue = 0 - value;
+    result = addToDate(date, DateTimeUnit::kDay, subValue);
+    return true;
```
Make this function return 'void' and remove this statement.

zwan5 replied:
Done

mbasmanova commented on velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -208,5 +214,42 @@ TEST_F(DateTimeFunctionsTest, lastDay) {
+    return evaluateOnce<Date>("date_sub(c0, c1)", date, value);
+  };
+
+  // Check null behaviors
```
Remove these tests. They test the framework, not the function.

zwan5 replied:
@mbasmanova Checked the Spark and Presto tests in DateTimeFunctionsTest.cpp, looks all used like this style. Can you check further. Thanks.

mbasmanova replied:
These test may need to be updated as well. Reviewers do not always catch everything.

zwan5 replied:
Done

mbasmanova commented on velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -208,5 +214,42 @@ TEST_F(DateTimeFunctionsTest, lastDay) {
+  EXPECT_EQ(std::nullopt, dateSubInt16(std::nullopt, 1));
+  EXPECT_EQ(std::nullopt, dateSubInt8(std::nullopt, 1));
+
+  // Simple tests
```
Make sure all comments are full sentences and end with a period.

zwan5 replied:
Done

mbasmanova commented on velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -208,5 +214,42 @@ TEST_F(DateTimeFunctionsTest, lastDay) {
+  EXPECT_EQ(parseDate("2019-02-28"), dateSubInt8(parseDate("2019-03-01"), 1));
+
+  // Account for the last day of a year-month
+  EXPECT_EQ(
```
Consider refactoring to make this shorter and easier to read:

```
EXPECT_EQ("2019-01-30", dateSubInt16("2020-02-29", 395);
```

zwan5 commented:
> @zwan5 Thank you for the contribution. A few rules to follow when adding functions:
> 
> 1- PR title should be "Add xxx Spark function" 2- PR description should not repeat PR title. Either leave empty or provide non-trivial description of the function behavior or implementation.

zwan5 commented:
@mbasmanova Thanks a lot for you review and constructive suggestions. I have reworked following the comments. can you help to review again? Thanks.

mbasmanova commented on velox/functions/sparksql/DateTimeFunctions.h:
```diff
@@ -209,4 +209,31 @@ struct LastDayFunction {
+struct DateSubFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void call(
```
Consider making this a template to avoid repeating the function.

```
template <typename T>
FOLLY_ALWAYS_INLINE void call(
      out_type<Date>& result,
      const arg_type<Date>& date,
      T value) {
    int32_t subValue = 0 - value;
    result = addToDate(date, DateTimeUnit::kDay, subValue);
  }
```

zwan5 replied:
Done

mbasmanova commented on velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -198,5 +202,47 @@ TEST_F(DateTimeFunctionsTest, lastDay) {
+        "date_sub(c0, c1)", {date, value}, {DATE(), INTEGER()});
+  };
+
+  // Check null behaviors.
```
Remove these.

mbasmanova commented on velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -198,5 +202,47 @@ TEST_F(DateTimeFunctionsTest, lastDay) {
+  };
+  const auto dateSubInt16 = [&](std::optional<int32_t> date,
+                                std::optional<int16_t> value) {
+    return evaluateOnce<int32_t, int32_t>(
```
Are these template parameters correct? This implementation looks exactly the same as dateSubInt32.

zwan5 replied:
The difference is on dateSub value type.

mbasmanova replied:
I see. It is confusing to see `evaluateOnce<int32_t, int32_t>` in all these lambdas. It makes me think that inputs are always int32_t. Maybe change to `evaluateOnce<int32_t, int32_t, int8_t>` or `evaluateOnce<int32_t>`.

zwan5 replied:
Before implementation registered 3 date_sub functions by different input value types, this may introduce confusion and duplication. I reworked and keep only one function. Kindly help to review again? Thanks.

mbasmanova reviewed (commented):
@zwan5 Thank you for iterating on the PR. Please, check PR description. It says exactly the same as title. Please, clear the description if there is nothing to add to the title.

zwan5 commented:
@mbasmanova Updated according to the comments. Can you help to review again? Thanks.

mbasmanova commented on velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -198,5 +202,28 @@ TEST_F(DateTimeFunctionsTest, lastDay) {
+  // Check simple tests.
+  int32_t inputDate = parseDate("2019-03-01");
+  int32_t expectedDate = parseDate("2019-02-28");
+  EXPECT_EQ(expectedDate, dateSub(inputDate, 1));
```
These tests might be easier to read as

```
EXPECT_EQ('2019-02-28', dateSub('2019-03-01', 1));
```

zwan5 replied:
Due to input parameters type limitation, didn't get exactly the same as  EXPECT_EQ('2019-02-28', dateSub('2019-03-01', 1)). Reworked as EXPECT_EQ(parseDate("2019-02-28"), dateSub("2019-03-01", 1)). Kindly help to review again. Thanks.

mbasmanova commented on velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -198,5 +202,28 @@ TEST_F(DateTimeFunctionsTest, lastDay) {
+  // Check for negative intervals.
+  inputDate = parseDate("2019-02-28");
+  expectedDate = parseDate("2020-02-29");
+  EXPECT_EQ(expectedDate, dateSub(inputDate, -366));
```
What happens if you pass very large positive/negative number as input?

zwan5 replied:
Added integer overflow check of int32_t value to DateSubFunction in DateTimeFunctions.h. Kindly help to review again. Thanks.

mbasmanova reviewed (commented):
@zwan5 Looks good % a couple of nits.

mbasmanova commented on velox/functions/sparksql/DateTimeFunctions.h:
```diff
@@ -209,4 +209,21 @@ struct LastDayFunction {
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Date>& result,
+      const arg_type<Date>& date,
+      const int64_t value) {
```
Why is this int64_t? Shouldn't it be int32_t? 

Please, clarify.

zwan5 replied:
Hi Basmanova，
Used int64_t to check the overflow of int32_t in case of user input unexpected very large value. 
But the actually valid input value is int32_t. 
See below function implementation. 
Kindly let me know if any comments. Thanks
    //check overflow
    if (value != (int32_t)value) {
      VELOX_UNSUPPORTED("integer overflow");
    }
	//actually valid int32_t input value
    int32_t subValue = 0 - (int32_t)value;
    result = addToDate(date, DateTimeUnit::kDay, subValue);

mbasmanova replied:
I think the signature should use int32_t, but the original question I had was about what happens if users pass std::numeric_limits<int32_t>::min/max? Do these work?

zwan5 replied:
Hi Basmanova，
Thanks a lot for your patient review.
I didn't understand your question initially. Yes, we should use int32_t. 
I have tried the min/max of <int32_t> test case.  min date is "-5877641-06-25", max date is "5881580-07-10".
min/max value -/+ 2147483647 works. Kindly refer to below logs.
I didn't add the max value test case yet, it's too huge. ^_^
date_sub spark function calls presto addToDate function, presto date_add also use it. if need to add protection, it's better to add in addToDate function.
Kindly let me know if any further comments. 
Have a good weekend.

//Below is boundary of the success test cases.
EXPECT_EQ(parseDate("5881580-07-10"), dateSub("1969-12-31", -2147483647));
EXPECT_EQ(parseDate("-5877641-06-25"), dateSub("1970-01-02", 2147483647));

//When exceed the boundary, below test case failed. 
 EXPECT_EQ(parseDate("-5877641-06-25"), dateSub("1969-01-01", 2147483647));
 EXPECT_EQ(parseDate("5881580-07-10"), dateSub("1970-12-31", -2147483647));
 ../../velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:231: Failure
Expected equality of these values:
  parseDate("-5877641-06-25")
    Which is: -2147483646
  dateSub("1969-01-01", 2147483647)
    Which is: (2147483284)
../../velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:232: Failure
Expected equality of these values:
  parseDate("5881580-07-10")
    Which is: 2147483646
  dateSub("1970-12-31", -2147483647)
    Which is: (-2147483285)

mbasmanova replied:
> I have tried the min/max of <int32_t> test case. min date is "-5877641-06-25", max date is "5881580-07-10".
min/max value -/+ 2147483647 works. Kindly refer to below logs.

Thank you for checking. Do these results match Spark?

mbasmanova replied:
> I didn't add the max value test case yet, it's too huge. ^_^

I'm not sure I understand. What is "too huge"? Any chance you could elaborate?

zwan5 replied:
Sorry for confusion , I mean the numbers is so huge. 
I am not sure if I should add these test cases? likes below.
What's your suggestion?
  EXPECT_EQ(parseDate("2019-02-28"), dateSub("2019-03-01", 1));
  EXPECT_EQ(parseDate("2019-01-30"), dateSub("2020-02-29", 395));
  EXPECT_EQ(parseDate("2020-02-29"), dateSub("2019-02-28", -366));
  EXPECT_EQ(parseDate("1970-01-02"), dateSub("-5877641-06-25", -2147483647));
  EXPECT_EQ(parseDate("1969-12-31"), dateSub("5881580-07-10", 2147483647));
  EXPECT_EQ(parseDate("5881580-07-10"), dateSub("1969-12-31", -2147483647));
  EXPECT_EQ(parseDate("-5877641-06-25"), dateSub("1970-01-02", 2147483647));

zwan5 replied:
Hi @mbasmanova,
Have confirmed the results matched with Spark. Test cased added also. Kindly help to review again. Thanks.

mbasmanova reviewed (approved):
@zwan5 Thanks.

mbasmanova commented:
Seeing failures when running with UBSAN:

```
[----------] 1 test from DateTimeFunctionsTest
[ RUN      ] DateTimeFunctionsTest.dateSub

buck-out/v2/gen/fbcode/9c7eecad4ba7475f/velox/functions/sparksql/__velox_functions_spark__/buck-headers/velox/functions/sparksql/DateTimeFunctions.h:220:26: runtime error: signed integer overflow: 0 - -2147483648 cannot be represented in type 'int'

SUMMARY: UndefinedBehaviorSanitizer: signed-integer-overflow buck-out/v2/gen/fbcode/9c7eecad4ba7475f/velox/functions/sparksql/__velox_functions_spark__/buck-headers/velox/functions/sparksql/DateTimeFunctions.h:220:26 in 
```

PHILO-HE commented:
Hi @mbasmanova, I'm just aware that the `date_add` function can be unified with this `date_sub` function. Is it proper to include `date_add` here? Or in a separate PR?

mbasmanova commented:
@zwan5 A separate PR might be better.

PHILO-HE commented:
> @zwan5 A separate PR might be better.

I will create a follow-up PR. Thanks!

zwan5 commented:
> Seeing failures when running with UBSAN:
> 
> ```
> [----------] 1 test from DateTimeFunctionsTest
> [ RUN      ] DateTimeFunctionsTest.dateSub
> 
> buck-out/v2/gen/fbcode/9c7eecad4ba7475f/velox/functions/sparksql/__velox_functions_spark__/buck-headers/velox/functions/sparksql/DateTimeFunctions.h:220:26: runtime error: signed integer overflow: 0 - -2147483648 cannot be represented in type 'int'
> 
> SUMMARY: UndefinedBehaviorSanitizer: signed-integer-overflow buck-out/v2/gen/fbcode/9c7eecad4ba7475f/velox/functions/sparksql/__velox_functions_spark__/buck-headers/velox/functions/sparksql/DateTimeFunctions.h:220:26 in 
> ```

Hi @mbasmanova,
How to run UBSAN test? Before I test using 'make unittest' and velox_functions_spark_test. Thanks

zwan5 commented on velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -198,5 +202,35 @@ TEST_F(DateTimeFunctionsTest, lastDay) {
+  };
+
+  // Check simple tests.
+  EXPECT_EQ(parseDate("2019-02-28"), dateSub("2019-03-01", 1));
```
Already added 0 days test case. Thanks

mbasmanova commented:
> I am not sure how to run UBSAN test?

@kgpai Krishna, would you have any instructions?

mbasmanova commented on velox/functions/sparksql/DateTimeFunctions.h:
```diff
@@ -209,4 +209,21 @@ struct LastDayFunction {
+      const int32_t value) {
+    int64_t subValue = 0 - (int64_t)value;
+    if (subValue != (int32_t)subValue) {
+      VELOX_UNSUPPORTED("integer overflow");
```
Does this behavior match Spark?

zwan5 replied:
Before implementation is wrong, treat kMin ( -2147483648) as overflow, this didn't match spark.
Update the function implementation to judge the corner case when input is kMin.
Kindly help to review. Thanks.

mbasmanova commented on velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -198,5 +202,39 @@ TEST_F(DateTimeFunctionsTest, lastDay) {
+  constexpr int32_t kMin = std::numeric_limits<int32_t>::min();
+  constexpr int32_t kMax = std::numeric_limits<int32_t>::max();
+  int32_t min = kMin + 1;
+  EXPECT_THROW(dateSub("-5877641-06-25", kMin), VeloxUserError);
```
use VELOX_ASSERT_THROW

mbasmanova reviewed (commented):
Some questions.

kgpai commented:
@mbasmanova Unfortunately Masha at the present moment trying to run UBSAN will require us to build and compile with LLVM. You can see details on why here : https://github.com/facebookincubator/velox/discussions/5131 . We are hoping to move all of our builds to use LLVM early this half and enable sanitizers by default. Unfortunately at the present moment this means its non trivial for external contributors to run UBSAN.

mbasmanova commented on velox/functions/sparksql/DateTimeFunctions.h:
```diff
@@ -209,4 +209,25 @@ struct LastDayFunction {
+      const arg_type<Date>& date,
+      const int32_t value) {
+    constexpr int32_t kMin = std::numeric_limits<int32_t>::min();
+    int32_t subValue = 0;
```
This variable is not used outside of if-else block. Please, move it inside if and else bodies or remove altogether.

mbasmanova commented on velox/functions/sparksql/DateTimeFunctions.h:
```diff
@@ -209,4 +209,25 @@ struct LastDayFunction {
+      subValue = 0 - value;
+      result = addToDate(date, DateTimeUnit::kDay, subValue);
+    } else {
+      subValue = 0 - (kMin + 1);
```
It is not obvious what's going on here. Please, add a comment.

mbasmanova reviewed (commented):
Thanks. Looks good % a couple of nits.

zwan5 commented on velox/functions/sparksql/DateTimeFunctions.h:
```diff
@@ -222,6 +222,8 @@ struct DateSubFunction {
       int32_t subValue = 0 - value;
       result = addToDate(date, DateTimeUnit::kDay, subValue);
     } else {
+      // If input value is minimum value kMin, use two steps to do DateSub.
```
Done. Thanks for the reminder.

mbasmanova commented on velox/functions/sparksql/DateTimeFunctions.h:
```diff
@@ -209,4 +209,26 @@ struct LastDayFunction {
+      int32_t subValue = 0 - value;
+      result = addToDate(date, DateTimeUnit::kDay, subValue);
+    } else {
+      // If input value is kMin, getting subValue(0-value) lead to overflow.
```
Thanks. Perhaps, rephrase for clarity.

```
If input values is kMin,  0 - value overflows. Subtract kMin in 2 steps to avoid overflow: - (- (kMin + 1)), then - 1.
```

I have a question though. if 'date' is some positive integer, adding kMin to it would overflow anyway, no? What happens when we call date_sub on 2023-07-10 and kMin?

zwan5 replied:
These comments are more clear. Thanks. 
When your mentioned cases happened, overflow happened. Velox code and Spark show the same results. Details see below logs:
spark-sql> SELECT date_sub('1969-12-31', 2147483647);
-5877641-06-23
spark-sql> SELECT date_sub('1969-12-30', 2147483647);
+5881580-07-11
spark-sql> SELECT date_sub('1969-12-31', -2147483648);
+5881580-07-11
spark-sql> SELECT date_sub('1970-01-01', -2147483648);
-5877641-06-23
spark-sql> SELECT date_sub('2023-07-10', -2147483648);
-5877588-12-29

mbasmanova replied:
> spark-sql> SELECT date_sub('2023-07-10', -2147483648);
> -5877588-12-29

Here, looks like we wrap around, not overflow. Is this how Spark behaves? Let's document this behavior and add this case to the test for illustration.

zwan5 replied:
yes, spark is the same behavior.

mbasmanova commented:
CI is red.

zwan5 commented:
> CI is red.
The CI test linux-adapters and linux-build got the same failure. Look have no relationship with this PR.
The following tests FAILED:
	267 - velox_exec_test (Child aborted)
Errors while running CTest
Exited with code exit status 8
CircleCI received exit code 8

mbasmanova commented:
@kgpai Looks like main may be broken? Would you help take a look?

zwan5 commented:
> @kgpai Looks like main may be broken? Would you help take a look?
CI run unit tests still failed, but I run the unit test success locally.
99% tests passed, 1 tests failed out of 256
Total Test time (real) = 1319.07 sec
The following tests did not run:
	 39 - */MemoryAllocatorTest.allocContiguousVsize/* (Disabled)
	 81 - */MemoryPoolTest.memoryLeakCheck/* (Disabled)
	 82 - */MemoryPoolTest.growBeyondMaxCapacity/* (Disabled)
	161 - SharedArbitrationTest.reclaimFromAggregation (Disabled)
The following tests FAILED:
	168 - SharedArbitrationTest.raceBetweenTaskTerminateAndReclaim (Child aborted)
Errors while running CTest
Exited with code exit status 8
CircleCI received exit code 8

zwan5 commented:
> @kgpai Looks like main may be broken? Would you help take a look?

Looks CI still failed, runnig unit tests failed, but I locally run the unit test successfully.
99% tests passed, 1 tests failed out of 256
Total Test time (real) = 1319.07 sec
The following tests did not run:
39 - /MemoryAllocatorTest.allocContiguousVsize/ (Disabled)
81 - /MemoryPoolTest.memoryLeakCheck/ (Disabled)
82 - /MemoryPoolTest.growBeyondMaxCapacity/ (Disabled)
161 - SharedArbitrationTest.reclaimFromAggregation (Disabled)
The following tests FAILED:
168 - SharedArbitrationTest.raceBetweenTaskTerminateAndReclaim (Child aborted)
Errors while running CTest
Exited with code exit status 8
CircleCI received exit code 8

zwan5 commented:
> CI is red.

hi @mbasmanova, looks CI issue is fixed now. Now CI test passed. Thanks

mbasmanova commented on velox/docs/functions/spark/datetime.rst:
```diff
@@ -7,6 +7,12 @@ Convenience Extraction Functions
+
+    Returns the date that is num_days before start_date.
+
+    num_days can be positive or negative.
```
Would you document the overflow / wrap-around behavior to avoid surprising users?

mbasmanova reviewed (commented):
@zwan5 Looks good % one nit for documentation.

mbasmanova commented:
@zwan5 Would you rebase so I could merge?

zwan5 commented:
> @zwan5 Would you rebase so I could merge?

hi @mbasmanova 
Have rebased. Thanks a lot.

mbasmanova commented:
@zwan5 Thank you for rebasing.

mbasmanova commented:
@zwan5 I'm seeing test failures:

```
[ RUN      ] DateTimeFunctionsTest.dateSub

fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/chrono:639:34: runtime error: signed integer overflow: 19548 + 2147483647 cannot be represented in type 'int'
    #0 0x7fa50833489f in std::common_type<std::chrono::duration<int, std::ratio<86400l, 1l> >, std::chrono::duration<int, std::ratio<86400l, 1l> > >::type std::chrono::operator+<int, std::ratio<86400l, 1l>, int, std::ratio<86400l, 1l> >(std::chrono::duration<int, std::ratio<86400l, 1l> > const&, std::chrono::duration<int, std::ratio<86400l, 1l> > const&) ./fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/chrono:639
    #1 0x7fa50833246a in std::chrono::time_point<std::chrono::_V2::system_clock, std::common_type<std::chrono::duration<int, std::ratio<86400l, 1l> >, std::chrono::duration<int, std::ratio<86400l, 1l> > >::type> std::chrono::operator+<std::chrono::_V2::system_clock, std::chrono::duration<int, std::ratio<86400l, 1l> >, int, std::ratio<86400l, 1l> >(std::chrono::time_point<std::chrono::_V2::system_clock, std::chrono::duration<int, std::ratio<86400l, 1l> > > const&, std::chrono::duration<int, std::ratio<86400l, 1l> > const&) ./fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/chrono:986
```

zwan5 commented:
> @zwan5 I'm seeing test failures:
> 
> ```
> [ RUN      ] DateTimeFunctionsTest.dateSub
> 
> fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/chrono:639:34: runtime error: signed integer overflow: 19548 + 2147483647 cannot be represented in type 'int'
>     #0 0x7fa50833489f in std::common_type<std::chrono::duration<int, std::ratio<86400l, 1l> >, std::chrono::duration<int, std::ratio<86400l, 1l> > >::type std::chrono::operator+<int, std::ratio<86400l, 1l>, int, std::ratio<86400l, 1l> >(std::chrono::duration<int, std::ratio<86400l, 1l> > const&, std::chrono::duration<int, std::ratio<86400l, 1l> > const&) ./fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/chrono:639
>     #1 0x7fa50833246a in std::chrono::time_point<std::chrono::_V2::system_clock, std::common_type<std::chrono::duration<int, std::ratio<86400l, 1l> >, std::chrono::duration<int, std::ratio<86400l, 1l> > >::type> std::chrono::operator+<std::chrono::_V2::system_clock, std::chrono::duration<int, std::ratio<86400l, 1l> >, int, std::ratio<86400l, 1l> >(std::chrono::time_point<std::chrono::_V2::system_clock, std::chrono::duration<int, std::ratio<86400l, 1l> > > const&, std::chrono::duration<int, std::ratio<86400l, 1l> > const&) ./fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/chrono:986
> 
> SUMMARY: UndefinedBehaviorSanitizer: signed-integer-overflow fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/chrono:639:34 in 
> ```

Hi @mbasmanova,
Below added test case lead to overflow.
    //EXPECT_EQ(parseDate("-5877588-12-29"), dateSub("2023-07-10", kMin));
DateSub called addToDate. See the code of addToDate, looks overflow handle is not implemented.
I suggest to remove the dateSub("2023-07-10", kMin)) testcase. How to you think about it?
//addToDate code
int32_t
addToDate(const int32_t input, const DateTimeUnit unit, const int32_t value) {
  // TODO(gaoge): Handle overflow and underflow with 64-bit representation

mbasmanova commented:
> I suggest to remove the dateSub("2023-07-10", kMin)) testcase. How to you think about it?

Removing a test case just makes the error go away, but the problem is still there waiting to show up in production. It would be preferable to fix the issue.

zwan5 commented:
> Hi @mbasmanova,

Hi @mbasmanova,
Yes, I can't agree more.
I have added the check of overflow. Also updated the documents. 
Kindly help to review again. 
Thanks.

zwan5 commented:
> > I suggest to remove the dateSub("2023-07-10", kMin)) testcase. How to you think about it?
> 
> Removing a test case just makes the error go away, but the problem is still there waiting to show up in production. It would be preferable to fix the issue.

hi @mbasmanova
About this case of dateSub("2023-07-10", kMin)), Spark didn't report overflow but return a negative date. should be a date > 5881580-07-11. So actually overflow happened. 
spark-sql> SELECT date_sub('1969-12-31', -2147483648);
+5881580-07-11
spark-sql> SELECT date_sub('2023-07-10', -2147483648);
-5877588-12-29
About the fix of this issue, I have two ideas.
1.  Add inputs overflow check of date and value, if (date - value) lead to int32 overflow,  throws an error. eg,  dateSub("2023-07-10", kMin)). But in this case, won't get the result -5877588-12-29 like Spark. 
    About this method,  I have push the implementation to github.
2.  Don't check overflow in cases like dateSub("2023-07-10", kMin)), return the same results as Spark.  But the testcase 
dateSub("2023-07-10", kMin)) will cause UBSAN overflow failure, because date_sub will lead to the calculation of '19548 + 2147483647'.  I didn't find ways to fix this error till now. 
Kindly let me what's 's your opinion about this issue? Thanks.

mbasmanova commented:
@zwan5 We need to match Spark's behavior. If we need to allow overflow, then, perhaps, silence UBSAN. See velox/functions/prestosql/ArithmeticImpl.h for examples.

```
__attribute__((__no_sanitize__("signed-integer-overflow")))
```

zwan5 commented:
> @zwan5 We need to match Spark's behavior. If we need to allow overflow, then, perhaps, silence UBSAN. See velox/functions/prestosql/ArithmeticImpl.h for examples.
> 
> ```
> __attribute__((__no_sanitize__("signed-integer-overflow")))
> ```

hi @mbasmanova ,
Thanks for reminder.
I have rollback the code to before version. And allow overflow as the example.
But I am not sure if it works, because external contributors not available for UBSAN test so far as Krishna mentioned before.
Kindly help to have a look. Thanks.

mbasmanova commented:
@zwan5 This still fails with UBSAN error. I don't have bandwidth to work on a proper fix right now. I suggest to create a GitHub issue to describe the desired behavior (Spark) and invite others to help figure out how to implement it.

zwan5 commented:
> @zwan5 This still fails with UBSAN error. I don't have bandwidth to work on a proper fix right now. I suggest to create a GitHub issue to describe the desired behavior (Spark) and invite others to help figure out how to implement it.

hi @mbasmanova 
Sorry for later response because I am taking annual leave last week.
Now we have confirmed the output of dateSub("2023-07-10", kMin)) matched with Spark behavior. The issue is failed to silence overflow check so UBSAN reported error, not a really failure. 
I am thinking if we can delete this test case in DateTimeFunctionsTest.cpp? See two reasons below :
1) dateSub("2023-07-10", kMin)) is a invalid corner case, it have exceed the valid date. I am not sure if somebody will actually use the date_sub function like this?
2) In order to pass the test case we need to silence overflow check, that will introduce additional efforts and redundant code. If it really valuable to do this? Looks there is no similar test cases in now Presto and Spark functions.
//EXPECT_EQ(parseDate("-5877588-12-29"), dateSub("2023-07-10", kMin));
Kindly let me know how do you think about it? Thanks

mbasmanova commented:
@zwan5 Sounds fine. Thanks. Let me try to merge.

mbasmanova commented:
@zwan5 Would you, please, rebase?

zwan5 commented:
> @zwan5 Would you, please, rebase?

Hi @mbasmanova  Have rebased. Thanks

zwan5 commented:
> @zwan5 Would you, please, rebase?

Hi @mbasmanova 
date_add merged, and conflict happened. Resolved the conflict and rebased.
Can you help to merge this PR? Thanks

mbasmanova commented:
@zwan5 Thank you for rebasing. Let me try to merge this PR.
